### PR TITLE
EX-335: RTCM encoders

### PR DESCRIPTION
**Pull request by nsirola**
Move the RTCM encoder code from tests to the public interface.

There might still be some gotchas when code-carrier divergence grows so large it does not fit into the bitfield, and especially on initializing and re-initializing the integer ambiguities, although this looks like to be already handled in ME:

> At start up and after each cycle slip, the initial ambiguity is reset and chosen so that the L1 Phaserange should match the L1 Pseudorange as closely as possible (i.e., within 1/2 L1 cycle) while not destroying the integer nature of the original carrier phase observation.

Note: MSM encoding to be implemented separately under EX-359

Tested with a 18 hour log of real data in https://github.com/swift-nav/gnss-converters/pull/120

### CODE REVIEW

File: c/src/encode.c [hunk 1]
```diff
@@ -286,14 +372,19 @@ uint16_t rtcm3_encode_1001(const rtcm_obs_message *msg_1001, uint8_t buff[]) {
       rtcm_setbitu(buff, bit, 6, msg_1001->sats[i].svId);
       bit += 6;
       encode_basic_freq_data(&msg_1001->sats[i].obs[L1_FREQ],
-                             GPS_L1_HZ,
+                             L1_FREQ,
                              &msg_1001->sats[i].obs[L1_FREQ].pseudorange,
                              buff,
                              &bit);
       ++num_sats;
+      if (num_sats >= RTCM_MAX_SATS) {
```
Comment (pmiettinen) [inline c/src/encode.c]: `if (++num_sats >= RTCM_MAX_SATS)`

File: c/src/encode.c [hunk 2]
```diff
@@ -334,11 +425,15 @@ uint16_t rtcm3_encode_1002(const rtcm_obs_message *msg_1002, uint8_t buff[]) {
...
       rtcm_setbitu(buff,
-              bit,
-              8,
-              (uint8_t)roundl(msg_1002->sats[i].obs[L1_FREQ].cnr * 4.0));
+                   bit,
+                   8,
+                   (uint8_t)round(msg_1002->sats[i].obs[L1_FREQ].cnr * 4.0));
       bit += 8;
       ++num_sats;
+      if (num_sats >= RTCM_MAX_SATS) {
```
Comment (pmiettinen) [inline c/src/encode.c]: same

File: c/src/encode.c [hunk 3]
```diff
@@ -403,22 +500,26 @@ uint16_t rtcm3_encode_1004(const rtcm_obs_message *msg_1004, uint8_t buff[]) {
...
       rtcm_setbitu(buff,
-              bit,
-              8,
-              (uint8_t)roundl(msg_1004->sats[i].obs[L2_FREQ].cnr * 4.0));
+                   bit,
+                   8,
+                   (uint8_t)round(msg_1004->sats[i].obs[L2_FREQ].cnr * 4.0));
       bit += 8;
       ++num_sats;
+      if (num_sats >= RTCM_MAX_SATS) {
```
Comment (pmiettinen) [inline c/src/encode.c]: same

File: c/src/encode.c [hunk 4]
```diff
@@ -542,9 +643,14 @@ uint16_t rtcm3_encode_1010(const rtcm_obs_message *msg_1010, uint8_t buff[]) {
 
       rtcm_setbitu(buff, bit, 7, amb);
       bit += 7;
-      rtcm_setbitu(buff, bit, 8, (uint8_t)roundl(sat_obs->obs[L1_FREQ].cnr * 4.0));
+      rtcm_setbitu(
+          buff, bit, 8, (uint8_t)round(sat_obs->obs[L1_FREQ].cnr * 4.0));
       bit += 8;
       ++num_sats;
+      if (num_sats >= RTCM_MAX_SATS) {
```
Comment (pmiettinen) [inline c/src/encode.c]: same

File: c/src/encode.c [hunk 5]
```diff
@@ -10,149 +10,235 @@
...
-  if (time < 168) return (time + 120) / 4;
-  if (time < 360) return (time + 408) / 8;
-  if (time < 744) return (time + 1176) / 16;
-  if (time < 937) return (time + 3096) / 32;
+static uint8_t to_lock_ind(double time) {
+  if (time < 24) {
+    return time;
+  }
+  if (time < 72) {
+    return (time + 24) / 2;
```
[the remark this answers is not in the document]
Comment (nsirola) [reply c/src/encode.c]: The correct logic here is to always truncate downwards.

File: c/src/encode.c (hunk 1 above)
Comment (pmiettinen) [reply c/src/encode.c]: I guess that's a fair argument. I am fine with both.

**Pull request merged**